File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        
        if(list1 == NULL)
            return list2;
        if(list2 == NULL)
            return list1;
        
        if(list1 -> val > list2 -> val) {
            ListNode* temp = list1;
            list1 = list2;
            list2 = temp;
        }
        ListNode* currhead = list1;
        while(list1 -> next != NULL and list1 -> next -> val <= list2 -> val) {
            list1 = list1 -> next; 
            cout<<list1->val<<endl;
        }
        
        ListNode* temp = list1 -> next;
        ListNode* temp2 = list2;
        list1 -> next = list2;
        if(temp2 -> next == NULL) {
            list2 -> next = temp;
            return currhead;
        }
        list2 -> next = mergeTwoLists(temp, temp2 -> next);
        
        return currhead;
        
    }
};

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        
        ch1, ch2 = list1, list2
        if list1.val > list2.val:
            ch1, ch2 = list2, list1
            
        first = ch1
        prev = None
        
        while ch1 and ch2:
            
            while ch1 and ch1.val <= ch2.val:
                prev = ch1
                ch1 = ch1.next
            
            prev.next = ch2
            ch2 = ch1
            ch1 = prev.next
        
        return first
            